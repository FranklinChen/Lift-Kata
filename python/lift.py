from enum import Enum

from attr import dataclass


class Direction(Enum):
    UP = 0
    DOWN = 1


class LiftSystem:
    def __init__(self, floors=None, calls=None, lifts=None):
        self.floors = list(floors) if floors else []
        self.calls = list(calls) if calls else []
        self.lifts = list(lifts) if lifts else []

    def reverse_order_floors(self):
        return reversed(self.floors)

    def calls_for(self, floor):
        return [c for c in self.calls if c.floor == floor]

    def lifts_on_floor(self, floor):
        return [l for l in self.lifts if l.floor == floor]


class Lift:

    def __init__(self, id, floor, doors_open=False, requested_floors=None):
        self.id = id
        self.floor = floor
        self.requested_floors = list(requested_floors) if requested_floors else []
        self.doors_open = doors_open


@dataclass
class Call:
    floor: int
    direction: Direction


