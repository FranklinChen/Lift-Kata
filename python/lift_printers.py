from lift import Direction


def print_lifts(lift_system):
    r = ""
    for floor in lift_system.reverse_order_floors():
        calls = " ".join([print_call(call) for call in (lift_system.calls_for(floor))])
        call_padding = whitespace(4-len(calls))
        lifts = " ".join([print_lift_for_floor(lift, floor) for lift in lift_system.lifts])
        floor_padding = whitespace(3 - len(str(floor)))
        r += f"{floor_padding}{floor} {calls}{call_padding} {lifts} .\n"

    return r


def print_lift_for_floor(lift, floor):
    if lift.floor == floor:
        return print_lift(lift)
    else:
        if floor in lift.requested_floors:
            return " * "
        else:
            return whitespace(3)


def print_lift(lift):
    if lift.doors_open:
        return f"]{lift.id}["
    else:
        return f"[{lift.id}]"


def print_call(call):
    direction = "^"
    if call.direction == Direction.DOWN:
        direction = "v"
    return f"{direction}"


def whitespace(size):
    w = ""
    for i in range(size):
        w += " "
    return w
