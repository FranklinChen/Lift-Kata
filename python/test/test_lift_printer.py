from approvaltests import verify

from lift import LiftSystem, Lift, Direction, Call
from lift_printers import print_lifts


def test_no_lifts():
    lift_system = LiftSystem(floors=[0, 1, 2, 3])
    verify(print_lifts(lift_system))


def test_one_lift():
    liftA = Lift("A", floor=0)
    lift_system = LiftSystem(floors=[0, 1, 2, 3], lifts=[liftA])
    verify(print_lifts(lift_system))


def test_one_lift_doors_open():
    liftA = Lift("A", floor=0, doors_open=True)
    lift_system = LiftSystem(floors=[0, 1, 2, 3], lifts=[liftA])
    verify(print_lifts(lift_system))


def test_several_lifts_with_requests():
    liftA = Lift("A", floor=0)
    liftA.requested_floors = [3]
    liftB = Lift("B", floor=2)
    lift_system = LiftSystem(floors=[0, 1, 2, 3], lifts=[liftA, liftB])
    verify(print_lifts(lift_system))


def test_several_lifts_with_requests_and_calls():
    liftA = Lift("A", floor=3)
    liftA.requested_floors = [0]
    liftB = Lift("B", floor=2)
    liftC = Lift("C", floor=2, doors_open=True)
    lift_system = LiftSystem(floors=[0, 1, 2, 3], lifts=[liftA, liftB, liftC], calls=[Call(1, Direction.DOWN)])
    verify(print_lifts(lift_system))


def test_large_lift_system():
    liftA = Lift("A", floor=0, requested_floors=[3,5,7])
    liftB = Lift("B", floor=2, doors_open=True)
    liftC = Lift("C", floor=-2)
    liftC.requested_floors = [0]
    liftD = Lift("D", floor=8, doors_open=True)
    liftD.requested_floors = [0, -1, -2]
    liftE = Lift("E", floor=10)
    liftE.requested_floors = [0,-1]
    lift_system = LiftSystem(floors=[-2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
                             lifts=[liftA, liftB, liftC, liftD, liftE],
                             calls=[Call(1, Direction.DOWN), Call(6, Direction.DOWN), Call(5, Direction.UP), Call(5, Direction.DOWN), Call(-1, Direction.UP)])
    verify(print_lifts(lift_system))
