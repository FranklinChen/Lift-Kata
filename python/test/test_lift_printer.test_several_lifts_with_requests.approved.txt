  3       *      .
  2          [B] .
  1              .
  0      [A]     .
